docs/spec/fastsync/fastsync.tla

```diff
@@ -0,0 +1,661 @@
+  state,                                     \* running or finished
+  blockPool
+  (*
+  blockPool [
+    height,                                 \* current height we are trying to sync. Last block executed is height - 1
+    peerIds,                                \* set of peers node is connected to
+    peerHeights,                            \* map of peer ids to its (stated) height
+    blockStore,                             \* map of heights to (received) blocks
+    receivedBlocks,                         \* map of heights to peer that has sent us the block (stored in blockStore)
+    pendingBlocks,                          \* map of heights to peer to which block request has been sent
+    syncHeight,                             \* height at the point syncTimeout was triggered last time
+    syncedBlocks                            \* number of blocks synced since last syncTimeout. If it is 0 when the next timeout occurs, then protocol terminates.
+  ]
+  *)
+
+\* the variables of the peers providing blocks
+VARIABLES
+  peersState
+  (*
+  peersState [
+    peerHeights,                             \* track peer heights
+    statusRequested,                         \* boolean set to true when StatusRequest is received. Models periodic sending of StatusRequests.
+    blocksRequested                          \* set of BlockRequests received that are not answered yet
+  ]
+  *)
+
+ \* the variables for the network and scheduler
+ VARIABLES
+  turn,                                     \* who is taking the turn: "Peers" or "Node"
+  inMsg,                                    \* a node receives message by this variable
+  outMsg                                    \* a node sends a message by this variable
+
+
+(* the variables of the node *)
+nvars == <<state, blockPool>>
+
+\* Control messages
+ControlMsgs ==
+    [type: {"addPeer"}, peerId: AllPeerIds]
+        \union
+    [type: {"removePeer"}, peerId: AllPeerIds]
+        \union
+    [type: {"syncTimeout"}]
+
+\* All messages (and events) received by a node
+InMsgs ==
+    {NoMsg}
+        \union
+    [type: {"blockResponse"}, peerId: AllPeerIds, block: BlocksWithNil]
+        \union
+    [type: {"statusResponse"}, peerId: AllPeerIds, height: Heights]
+        \union
+    ControlMsgs
+
+\* Messages sent by a node and received by peers (environment in our case)
+OutMsgs ==
+    {NoMsg}
+        \union
+    [type: {"statusRequest"}]           \* StatusRequest is broadcast to the set of connected peers.
+        \union
+    [type: {"blockRequest"}, peerId: AllPeerIds, height: Heights]
+
+
+(********************************** NODE ***********************************)
+
+InitNode ==
+     \E pIds \in SUBSET AllPeerIds \ {{}}:                   \* set of peers node established initial connections with
+        /\ blockPool = [
+                height |-> 1,
+                peerIds |-> pIds,
+                peerHeights |-> [p \in AllPeerIds |-> NilHeight],     \* no peer height is known
```

Comment:
It's a function definition (it's a domain of elements), and p can therefore be any element of the AllPeerIds. In case we would constrain it to peerIds we would need to change the function definition once we add a new element. In this case we initialise the same way elements that are in pIds and those that are not, but this doesn't necessarily to be always the case. For example, if we would start fastsync with set of peerIds for which we know heights, you would define this function differently for pIds, while others would be NilHeight.